### ISSUE TITLE
Reciprocal of complex 0 and `1 / numpy.array(0+0j)` give different results

### ISSUE THREAD

**Issue by kurtamohler** [numpy/numpy#17425]
`numpy.reciprocal(a)` and `1 / a` give different results with `a = numpy.array(0 + 0j)`. `reciprocal` gives `nan+nanj`, and `1 / a` gives `inf+nanj`.

It seems like neither of these results is correct, but I could certainly be wrong. I feel like the result in both cases should be `inf+0j`, since we get `inf` when we use real numbers: `1 / numpy.array(0.0)` and `numpy.reciprocal(numpy.array(0.0))`.

### Reproducing code example:

```
>>> import numpy

>>> a = numpy.array(0 + 0j)
__main__:1: RuntimeWarning: invalid value encountered in reciprocal
>>> numpy.reciprocal(a)
(nan+nanj)
>>> 1 / a
(inf+nanj)

>>> 1 / numpy.array(0.0)
inf
>>> numpy.reciprocal(numpy.array(0.0))
__main__:1: RuntimeWarning: divide by zero encountered in reciprocal
inf
```

### NumPy/Python version information:

```
>>> print(numpy.__version__, sys.version)
1.18.1 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
```

### OS

Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic

**Comment by eric-wieser** [comment]
>I feel like the result in both cases should be `inf+0j`,

What does the IEEE-754 spec say is the correct result?

**Comment by kurtamohler** [comment]
I'm not sure what the spec itself says--evidently I'd have to pay for it to find out. I don't see any mention of complex numbers on the wikipedia page, so maybe it's not covered: https://en.wikipedia.org/wiki/IEEE_754

For what it's worth, my version of g++ claims to support IEEE-754 for floats, and I get `1.0f / 0.0f = inf`, which agrees with numpy behavior in the case of real numbers.

**Comment by eric-wieser** [comment]
Right, my question is about what `g++` does for complex division. The C99 spec seems to say:

> if the first operand is a nonzero finite number or an infinity and the second operand is a zero, then the result of the / operator is an infinity.

and

> A complex or imaginary value with at least one infinite part is regarded as an infinity
(even if its other part is a NaN)

which suggests that the behavior of `1 / a` is correct, and `reciprocal` is not.

**Comment by pv** [comment]
C99 Annex G.5.1.4: the result from np.reciprocal is wrong, and the result from `1/np.complex_(0)` is correct. One should have `assert np.isinf(np.reciprocal(0+0j))`, but it's not specified which complex infinity is returned.

There's also something else strange here:
```
>>> z = np.complex_(0+0j)
>>> 1/z
(inf+nanj)
>>> 1/(1/z)
(nan+nanj)
```
Python doesn't consider `(inf, nan)` as an infinity:
```
>>> 1/complex(float("inf"), float("nan"))
(nan+nanj)
```
so it seems there's deviation from C99 spec also there.

**Comment by kurtamohler** [comment]
Ah, that makes sense. Thanks! I can try to make a PR to change `reciprocal` to match `1 / a` this weekend.

**Comment by skirpichev** [comment]
> Python doesn't consider (inf, nan) as an infinity in division

@pv, now it's a bug: https://github.com/python/cpython/issues/119372.  Here is a corresponding NumPy issue: https://github.com/numpy/numpy/issues/26560

**Comment by ikrommyd** [comment]
So this is in the direction of https://github.com/numpy/numpy/pull/30806 which we decided to not do now but not completely. The reason you are seeing different results in division vs reciprocal is what Eric said earlier above. It's not that division has full Annex G recoveries but because division has a special branch only when you divide by zero:
https://github.com/numpy/numpy/blob/7643e77e8481c80472efb1c9d29de9898d765187/numpy/_core/src/umath/loops.c.src#L2217-L2221
while reciprocal does not
https://github.com/numpy/numpy/blob/7643e77e8481c80472efb1c9d29de9898d765187/numpy/_core/src/umath/loops.c.src#L2329-L2347

@seberg - do you think perhaps we shouldn't close this one like the others given the Annex G discussion but add such an if statement only for division by zero in reciprocal too? This was added in 2011 for divide by in commit: https://github.com/numpy/numpy/commit/65b77ee94131bf8365d8a6dba6fa19da1269339c
An full Annex G correction automatically gets you the fix for this, but I'm only talking about a divide by zero if branch in the loop here.
